#320 - Unit Testing Style Guide (DRAFT) (#321)
Adobe-Marketing-Cloud/asset-share-commons: core/src/test/java/com/adobe/aem/commons/assetshare/README.md

# Unit Testing Guide

## Testing Tools

* Unit tests are broken up by the class being tested under `core/src/test/java`.

> * Unit tests are written using the dependency provides in the core/pom.xml:

Reviewer: typo:
```suggestion
* Unit tests are written using the dependency provided in the core/pom.xml:
```

    * [JUnit 4.12](https://junit.org/junit4/)
    * [Mockito 2](https://site.mockito.org/)
    * [AEM Mocks](http://wcm.io/testing/aem-mock/usage.html)

## Test Cases

> Each class being test has a corresponding test case class, sharing the package and filename of the class being test, however the filename is post-fixed with "Test".

Reviewer: typo:
```suggestion
Each class being tested has a corresponding test case class, sharing the package and filename of the class being tested, however the filename is post-fixed with "Test".
```

For example, to test:

   `core/src/main/java/com.adobe.aem.commons.assetshare.components.impl.HelloWorldImpl.java`

> 

Reviewer 2: you probably mean ```com/adobe/aem/commons/assetshare/components/impl/HelloWorldImpl.java```

this test is created:

    `core/src/test/java/com.adobe.aem.commons.assetshare.components.impl.HelloWorldImplTest.java`

### Test Methods

> Each public method on the class being test should have at least one corresponding test method. If multiple conditions (working/not-working) required testing, multiple test methods should be defined.

Reviewer: ```suggestion
Each public method on the class being tested should have at least one corresponding test method. If multiple conditions (working/not-working) required testing, multiple test methods should be defined.
```

Test method naming:
* Shares the name of method being tested, ex: `getTitle()`
* Prefixed with `test`, ex: the test method is `testGetTitle()`
* Optionally post-fixed with an underscore-prefixed, capitalized condition name, ex: `testGetTitle__EmptyTitle()`

post-fixed with a clear identifier `_ConditionToTest`

For example, to test in `HelloWorld.java`:

   `String getTitle()`

these test methods created in `HelloWorldImplTest.java`:

    `@Test void testGetTitle()` which tests for the normal, populated title condition
    `@Test void testGetTitle_EmptyTitle()`, which tests for the condition where the title is missing

> 

Reviewer: I like to name my test methods not only with the name of the method that they are testing, but also the behavior that they are testing.  For example, testGetTitle() would be testGetTitle_ReturnsSetTitle() or testGetTitle_EmptyTitle_ReturnsEmptyString().

Author: Lets start with title and condition; I think baking in the expected result into the method name might get super verbose. Ideally, if the tests are well writing it should be quite easy to derive what the expected result should be by reading the failing test.

### Resource files (JSON)

> AEM Mocks provide support to load JSON into mock AEM context and write tests against that, avoiding manual mocking of resources, OSGi services, etc. using AEM-agnostic tools like mockito.

Reviewer: ```suggestion
AEM Mocks provide support to load JSON into mock AEM context and write tests against that, avoiding manual mocking of resources, OSGi services, etc. using AEM-agnostic tools like Mockito.
```

The following convention for defining JSON resource files should be followed:

* JSON files are created under test/resources under a folder structure that follows the test case class's package and file name, for example:

    `core/src/test/resources/com/adobe/aem/commons/assetshare/components/impl/HelloWorldImplTest.json`

provides the JSON required by the test case `core/src/test/java/com.adobe.aem.commons.assetshare.components.impl.HelloWorldImplTest.java`.

> Each test case's JSON file can define multiple well-named root in the JSON definition that represent different states, which in tern can be used across 1 or many test methods in the corresponding test case.

Reviewer 2: typo: "in turn"

> Each test case's JSON file can define multiple well-named root in the JSON definition that represent different states, which in turn can be used across 1 or many test methods in the corresponding test case.

Reviewer: ```suggestion
Each test case's JSON file can define multiple, well-named JSON definition roots that represent different states and can in turn be used across one or many test methods in the corresponding test case.
```

For example, the HelloWorldImplTest.java may need to test for: a fully configured ("default")  component, an wholly un-configured component ("empty") and a component with ONLY the title populated ("title") using the JSON file defined below:

```
{
  "default": {
      jcr:primaryType: "nt:unstructured",
      sling:resourceType: "asset-share-commons/components/search/hello-world"
      title: "hello world!",
      propX: "foo"
  },
  "empty": {
      jcr:primaryType: "nt:unstructured",
      sling:resourceType: "asset-share-commons/components/search/hello-world"
  },
  "title": {
      jcr:primaryType: "nt:unstructured",
      sling:resourceType: "asset-share-commons/components/search/hello-world"
      title: "hello world!",
  }
```

####

```
@Before
public void setup() {

>    ctx.load().json("HelloWorldImplTest.json");

Author: Have to verify that this json will auto-resolve since its in the corresponding resources "package"

Reviewer 2: There is no ctx.json() method with 1 argument only, you always have to provide the path; and "/" as path throws an exception...

}

...

@Test
public void testGetTitle__EmptyTitle()  {

>     context.currentResource("/empty");

Author: yeh - reading thought this makes sense to me and not much extra to type

For example, the HelloWorldImplTest.java may need to test for: a fully configured ("default")  component, a wholly un-configured component ("empty") and a component with ONLY the title populated ("title") using the JSON file defined below:

```
{
  "default": {
      jcr:primaryType: "nt:unstructured",
      sling:resourceType: "asset-share-commons/components/search/hello-world"
      title: "hello world!",
      propX: "foo"
  },
  "empty": {
      jcr:primaryType: "nt:unstructured",
      sling:resourceType: "asset-share-commons/components/search/hello-world"
  },
  "title": {
      jcr:primaryType: "nt:unstructured",
      sling:resourceType: "asset-share-commons/components/search/hello-world"
      title: "hello world!",
  }
```

#### Example consumption of JSON

```
@Before
public void setup() {
   ctx.load().json("HelloWorldImplTest.json");
}

...

@Test
public void testGetTitle__EmptyTitle()  {
    context.currentResource("/empty");
    ...
}
```

## A sample is worth 1000 words

### Sample test case

`core/src/test/java/com/adobe/aem/commons/assetshare/components/impl/HelloWorldImplTest.java`

> ```

Reviewer 3: Do we want to add the package imports to the sample? Might help with potential same-name collisions...?
